File: ocaml/lib/parser.ml
open Maybe

type 'a parser = P of (string -> ('a * string) maybe)

let parse (P unp) inp = unp inp

let pure x = P (fun inp -> just (x, inp))

let (>>=) p f = P (fun inp -> (parse p inp) >>= (fun (x, out) -> parse (f x) out))

let bind p f = p >>= f

let (<$>) f p = p >>= fun a -> pure(f a)

let fmap f p = f <$> p

let (<*>) fp p = fp >>= (fun f -> f <$> p ())

let apply fp p = fp <*> p

let (>>) p fq = p >>= (fun _ -> fq ())
let (>>>) p q = p >>= (fun _ -> q)

let skip p q = p >>> q

let empty = P (fun _ -> nothing)

let (<|>) p q = P (fun inp -> parse p inp <|> (fun () -> parse (q ()) inp))
let (<||>) p q = p <|> (fun () -> q)

let or_else p q = p <||> q

File: R/src/parser.R
library(maybe)

as_parser <- function(.f) {
  structure(.f, class = "parser")
}

is_parser <- function(.p) {
  identical(class(.p), "parser")
}

assert_is_parser <- function(p) {
  if (is_parser(p))
    p
  else {
    print(class(p))
    stop(
      "The argument to a function must be a parser.",
      call. = FALSE
    )
  }
}

parse <- function(.p, inp) {
  (.p %>% assert_is_parser())(inp)
}

parser_map <- function(.p, .f){
  .p %>% assert_is_parser()
  as_parser(function(inp) { .p %>% parse(inp) %>% fmap(function (pair) { list(.f(pair[[1]]), pair[[2]]) }) })
}

parser_and_then <- function(.p, .f){
  .p %>% assert_is_parser()
  as_parser(function(inp) { .p %>% parse(inp) %>% and_then(function (pair) { .f(pair[[1]]) %>% parse(pair[[2]])  }) })
}

parser_skip <- function(.p, .fp){
  .p %>% parser_and_then(function(x) { .fp() })
}

parser_apply <- function(.pf, .fp){
  .pf %>% parser_and_then(function(f) { .fp() %>% parser_map(f) })
}

parser_or_else <- function(.p, .fp){
  .p %>% assert_is_parser()
  as_parser(function(inp) {
    m <- .p %>% parse(inp)
    if(is_just(m)) m else .fp() %>% parse(inp)
  })
}

pure <- function(x) {
  as_parser(function(inp){ just(list(x, inp)) })
}

empty <- as_parser(function(inp){ nothing() })
